Implementation/Composing/index-v9.md

```diff
@@ -0,0 +1,519 @@
+        // spotify:track:7aCk4XfXIEJM2MecU6Gmf2
+
+        public override string[] UrlSchemeRegex => new string[]
+        {
+            @".*.spotify.com/.*",
+            @"spotify:.*"
+        };
+
+        public override Dictionary<string, string> RequestParams => new Dictionary<string, string>();
+
+        public override string GetMarkup(string url, int maxWidth = 0, int maxHeight = 0)
+        {
+            var requestUrl = base.GetEmbedProviderUrl(url, maxWidth, maxHeight);
+            var oembed = base.GetJsonResponse<OEmbedResponse>(requestUrl);
+
+            return oembed.GetHtml();
+        }
+    }
+}
+```
+
+```csharp
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Core.DependencyInjection;
+
+namespace My.Website
+{
+    public class CustomOEmbedComposer : IComposer
+    {
+        public void Compose(IUmbracoBuilder builder)
+        {
+            // Change the OEmbedProviders collection
+            // by adding our new EmbedProvider for Spotify
+            builder.OEmbedProviders().Append<Spotify>();
+        }
+    }
+}
+```
+See a list of collections below to determine which are 'type scanned' and which require explicit registration.
+
+### Example - Creating a Composer to listen for ContentSavingNotification
+
+This example shows how to create a component and an notification handler for the `ContentSavingNotification`, (perhaps to check for explicit words, or some custom business logic that needs to run before the content item is saved in Umbraco).
+
+We create a new C# class that implements `IComposer` and use it register our notification handler.
+
+
+```csharp
+using System.Linq;
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Core.DependencyInjection;
+using Umbraco.Cms.Core.Events;
+using Umbraco.Cms.Core.Notifications;
+using Umbraco.Extensions;
+
+namespace My.Website
+{
+    public class SubscribeToContentServiceSavingComposer : IComposer
+    {
+        public void Compose(IUmbracoBuilder builder)
+        {
+            builder.AddNotificationHandler<ContentSavingNotification, CustomContentSavingNotificationHandler>()
+        }
+    }
+
+    public class CustomContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
+    {
+        public void Handle(ContentSavingNotification notification)
+        {
+            foreach (var content in notification.SavedEntities
+                // Check if the content item type has a specific alias
+                .Where(c => c.ContentType.Alias.InvariantEquals("MyContentType")))
+            {
+                // Do something if the content is using the MyContentType doctype
+            }
+        }
+    }
+}
+```
+
+:::warning
+Ordering of composers is important, the last one added can override a previously added composer! Make sure, when overriding,
+that your composer that is doing the overriding, is 'composing', after the composer has 'composed' the element you wish to override!
+:::
```

Comment:
```suggestion
### Example - Creating a Composer to listen for ContentSavingNotification

This example shows how to create a component and an notification handler for the `ContentSavingNotification`, (perhaps to check for explicit words, or some custom business logic that needs to run before the content item is saved in Umbraco).

We create a new C# class that implements `IComposer` and use it register our notification handler.


```csharp
using System.Linq;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.DependencyInjection;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Notifications;
using Umbraco.Extensions;

namespace My.Website
{
    public class SubscribeToContentServiceSavingComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            builder.AddNotificationHandler<ContentSavingNotification, CustomContentSavingNotificationHandler>()
        }
    }

    public class CustomContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
    {
        public void Handle(ContentSavingNotification notification)
        {
            foreach (var content in notification.SavedEntities
                // Check if the content item type has a specific alias
                .Where(c => c.ContentType.Alias.InvariantEquals("MyContentType")))
            {
                // Do something if the content is using the MyContentType doctype
            }
        }
    }
}
```

:::warning
Ordering of composers is important, the last one added can override a previously added composer! Make sure, when overriding,
that your composer that is doing the overriding, is 'composing', after the composer has 'composed' the element you wish to override!
:::

### Example - Explicitly Registering a new custom OEmbedProvider
This example shows a custom 'Spotify' OEmbed Provider which will allow Spotify URLs to be used via the 'embed' button in the Grid and Rich Text Editors. As the collection for OEmbedProviders is not 'typed scanned', we need to explicitly register the provider in the collection of OEmbedProviders. We create a C# class which implements `IUserComposer` and append our new Spotify OEmbedProvider to the OEmbedProviders() collection:

```csharp

using System.Collections.Generic;
using Umbraco.Cms.Core.Media.EmbedProviders;
using Umbraco.Cms.Core.Serialization;

namespace My.Website
{
    public class Spotify : EmbedProviderBase
    {
        public Spotify(IJsonSerializer jsonSerializer)
            : base(jsonSerializer)
        {
        }

        public override string ApiEndpoint => "https://embed.spotify.com/oembed/";

        // Playlist
        // https://open.spotify.com/user/spotify/playlist/37i9dQZF1E4sNI4jZloSZr?si=cueBooBfTnqCGriSa4N_Kg
        // spotify:user:spotify:playlist:37i9dQZF1E4sNI4jZloSZr

        // Artist
        // https://open.spotify.com/artist/0iirUbtgwt9jEkc2Grin8C?si=TLeUR2cHR-KPRJJhW6YiVg
        // spotify:artist:0iirUbtgwt9jEkc2Grin8C

        // Album
        // https://open.spotify.com/album/0lvtdqkqIln6uDBBUT7DHL?si=XTVJIEmnS_OVv9l6ktPFiw
        // spotify:album:0lvtdqkqIln6uDBBUT7DHL

        // Track
        // https://open.spotify.com/track/7aCk4XfXIEJM2MecU6Gmf2?si=vESDzI0xTNeA9FQ_dvf1eQ
        // spotify:track:7aCk4XfXIEJM2MecU6Gmf2

        public override string[] UrlSchemeRegex => new string[]
        {
            @".*.spotify.com/.*",
            @"spotify:.*"
        };

        public override Dictionary<string, string> RequestParams => new Dictionary<string, string>();

        public override string GetMarkup(string url, int maxWidth = 0, int maxHeight = 0)
        {
            var requestUrl = base.GetEmbedProviderUrl(url, maxWidth, maxHeight);
            var oembed = base.GetJsonResponse<OEmbedResponse>(requestUrl);

            return oembed.GetHtml();
        }
    }
}
```

```csharp
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.DependencyInjection;

namespace My.Website
{
    public class CustomOEmbedComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            // Change the OEmbedProviders collection
            // by adding our new EmbedProvider for Spotify
            builder.OEmbedProviders().Append<Spotify>();
        }
    }
}
```
See a list of collections below to determine which are 'type scanned' and which require explicit registration.
```